arch/x86/kvm/pkvm/pkvm.c

```diff
@@ -508,6 +508,55 @@ static int pkvm_vcpu_load(int vm_handle, int vcpu_handle)
 	return ret;
 }
 
+static int pkvm_vcpu_put(int vm_handle, int vcpu_handle)
+{
+	struct pkvm_vcpu *pkvm_vcpu = pkvm_get_vcpu(vm_handle, vcpu_handle);
+	int cpu = raw_smp_processor_id(), loaded_cpu, ret = 0;
+	struct kvm_vcpu *vcpu;
+
+	if (!pkvm_vcpu)
+		return -EINVAL;
+
+	vcpu = &pkvm_vcpu->vcpu;
+	loaded_cpu = READ_ONCE(vcpu->cpu);
+	if (loaded_cpu == cpu) {
+		/*
+		 * The current active vCPU is the host vCPU. Switch to the guest
+		 * vCPU in case vcpu_put operation requires.
+		 */
+		kvm_x86_call(vcpu_load)(vcpu, cpu);
+
+		/*
+		 * Another guest vCPU may have already been loaded on this CPU
+		 * thus the cur_guest_vcpu may be overridden. So only set the
+		 * cur_guest_vcpu as NULL if it points to the guest vCPU being
+		 * put.
+		 */
+		if (vcpu == this_cpu_read(cur_guest_vcpu))
+			this_cpu_write(cur_guest_vcpu, NULL);
+
+		kvm_x86_call(vcpu_put)(vcpu);
+
+		/*
+		 * Put this pkvm_vcpu to allow it to be freed via the vcpu_free PV
+		 * interface.
+		 */
+		pkvm_put_vcpu(pkvm_vcpu);
+
+		/* Switch to the host vCPU as a guest vCPU was just loaded. */
+		kvm_x86_call(vcpu_load)(this_cpu_read(host_vcpu), cpu);
+
+		WRITE_ONCE(vcpu->cpu, -1);
+	} else if (loaded_cpu != -1) {
+		/* The guest vCPU was not loaded on this CPU. */
+		ret = -EBUSY;
```

Comment:
nit: so it was not necessarily busy, it was just not loaded on this CPU (and maybe not on any other CPU either). So maybe something like -EINVAL would be more suitable?